### PR TITLE
Add data to the service execute with no file

### PR DESCRIPTION
**Pull request by krhubert**
related to https://github.com/mesg-foundation/core/issues/256

### CODE REVIEW

**Comment by NicolasMahe** [comment]
@krhubert can you please reference the linked issue in the description at least

File: cmd/service/execute.go [hunk 1]
```diff
@@ -96,14 +101,28 @@ func getTaskKey(cmd *cobra.Command, serviceID string) string {
...
-		}
+func getData(cmd *cobra.Command) (string, error) {
+	data := cmd.Flag("data").Value.String()
+	jsonStr := cmd.Flag("json").Value.String()
+	jsonFile := cmd.Flag("json-file").Value.String()
+
+	if (data != "" && jsonStr != "") ||
+		(data != "" && jsonFile != "") ||
+		(jsonStr != "" && jsonFile != "") {
+		return "", errors.New("data, json and json-file flags are  mutually exclusive")
```
Comment (NicolasMahe) [inline cmd/service/execute.go]: Can you write something more user friendly like: "You can use only of one of the following flags: 'data', 'json' and 'json-file'."

**Comment by antho1404** [comment]
This one will definitely need some update on the documentation and changelog and also I'm not really sure about it first we already have a `json` that reads a file and not it does something different. Also I don't see the point of having a json directly in the command (not in the file) and having a data.

File: cmd/service/execute.go (hunk 1 above)
Comment (krhubert) [reply cmd/service/execute.go]: Sure, done.

As xpflag is eXtension to [spf13/pflag](https://github.com/spf13/pflag) that contains custom flags (implements [pflag.Value interface](https://godoc.org/github.com/spf13/pflag#Value)) and I put it in different package.

**Comment by krhubert** [comment]
@antho1404 For sure, I've kept 3 options for further discussion (which of them should/shouldn't stay).
As for JSON in command, I find it frustrating when I wanted to execute some task I had to create a file with simple JSON, then edit the file in order to execute with different task data, etc...

It's nice to have json in cmd:
```
$ mesg-core service test --task execute --data '{"key":"1"}'
$ mesg-core service test --task execute --data '{"key":"2"}'
$ # do more 
```

**Comment by antho1404** [comment]
I was actually against json as parameter because if using double quotes it looks like 
`"{\"foo\":\"bar\"}"` which is terrible to write but with single quotes it's way better. I'm just afraid for a user perspective they might do this mistake and spend hours to understand why their data are not good. Users might be junior developers so probably not that much comfortable with the terminal.

I like the data part but this will be interpreted as string and if the data needs to be boolean for example the core will reject the task so this is something to manage too

File: cmd/service/execute.go [hunk 2]
```diff
@@ -11,11 +13,13 @@ import (
 	"github.com/mesg-foundation/core/api/core"
 	"github.com/mesg-foundation/core/cmd/utils"
 	"github.com/mesg-foundation/core/service"
+	"github.com/mesg-foundation/core/utils/xpflag"
 	"github.com/spf13/cobra"
 	survey "gopkg.in/AlecAivazis/survey.v1"
 )
 
 var executions []*core.ResultData
```
Comment (NicolasMahe) [inline cmd/service/execute.go]: This is not used anymore.

**Comment by NicolasMahe** [comment]
@mesg-foundation/core What about asking the value of the task's inputs by questions if no data has been passed by flags? It's not the most efficient way to pass data, but for a quick test of a task it's super convenient! No need to know the inputs before executing the command.

File: cmd/service/execute.go (hunk 2 above)
Comment (krhubert) [reply cmd/service/execute.go]: don't get it, what do you mean? `executions` wasn't touched

**Comment by NicolasMahe** [comment]
To sum up the conversion on discord:

- [ ] Revert `--json` flag to accept a JSON file
- [ ] Keep flag `--data`. But be compatible with: `--data key1=value1 --data key2=value2`
- [ ] Cast every value passed with the `data` flag using the type defined in the definition of the service (https://docs.mesg.com/guide/service/service-file.html#service-file-2, https://github.com/mesg-foundation/service-ethereum/blob/master/mesg.yml)

**Comment by antho1404** [comment]
I would add also to update the documentation
https://github.com/mesg-foundation/core/blob/fc75e9b86e16947a351f0202ea72a16c6216941f/docs/guide/service/test.md

File: service/cast.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	task, ok := s.Tasks[taskKey]
+	if !ok {
+		return nil, fmt.Errorf("Task %q not found", taskKey)
+	}
+
+	m := make(map[string]interface{}, len(taskData))
+	for key, value := range taskData {
+		inputType, ok := task.Inputs[key]
+		if !ok {
+			return nil, fmt.Errorf("Task %q has no input named %q", taskKey, key)
```
Comment (NicolasMahe) [inline service/cast.go]: Please create a new type for this error in `service/error.go`

File: service/cast.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+				m[key] = i
+			}
+		case "Boolean":
+			b, err := strconv.ParseBool(value)
+			if err != nil {
+				return nil, fmt.Errorf("Task %q - input %q is not a Boolean type", taskKey, key)
+			}
+			m[key] = b
+		case "Object":
+			return nil, fmt.Errorf("Task %q - input %q - Object type is not supported", taskKey, key)
```
Comment (NicolasMahe) [inline service/cast.go]: Can we really not pass any Object with the CLI? Even a JSON stringified?

File: service/cast.go (hunk 3 above)
Comment (krhubert) [reply service/cast.go]: done

File: service/cast.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import (
+	"fmt"
+	"strconv"
+)
+
+// Cast converts map[string]string to map[string]interface{} based on defined types in the service tasks map.
+func (s *Service) Cast(taskKey string, taskData map[string]string) (map[string]interface{}, error) {
+	task, ok := s.Tasks[taskKey]
+	if !ok {
+		return nil, fmt.Errorf("Task %q not found", taskKey)
```
[the remark this answers is not in the document]
Comment (krhubert) [reply service/cast.go]: done

File: service/cast.go (hunk 4 above)
Comment (krhubert) [reply service/cast.go]: Ok, so I'll continue here instead of return an error

File: service/cast.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+func castBoolean(value string) (interface{}, error) {
+	b, err := strconv.ParseBool(value)
+	if err != nil {
+		return nil, fmt.Errorf("input %q is not a Boolean type", value)
+	}
+	return b, nil
+}
+func castObject(value string) (interface{}, error) {
+	// do not cast object type
+	return nil, nil
```
Comment (antho1404) [inline service/cast.go]: This is not returning any value so any service with object will fail.

We can add a basic support by at least returning the value or even better trying to parse it from string

**Pull request merged**